### DISCUSSION TITLE
tar_quarto rendering issues

### DISCUSSION THREAD

**Discussion by tklebel** [ropensci/tarchetypes#124] [Help] [answered]
I have a weird issue that I have not really been able to reproduce with a minimal example.

I have a fairly simple targets setup. This is my `_targets.R`:

```r
library(targets)
library(tarchetypes)

source("R/functions.R")

options(tidyverse.quiet = TRUE)
tar_option_set(packages = c("scales", "tidyverse", "hrbrthemes"))

list(
  # baseline file -----------
  tar_target(
    baseline_file,
    "../outputs/xyz.csv",
    format = "file"
  ),
  tar_target(
    baseline,
    read_experiments(baseline_file),
    format = "feather"
  ),
  tar_target(
    baseline_selection,
    select_baseline(baseline),
    format = "feather"
  ),
  tar_render(baseline_report, "01-analyse-baseline.qmd")

)
```

This "works" but obviously does not create a quarto styled doc but HTML in the style of RMarkdown. If I replace the last line with `  tarchetypes::tar_quarto(baseline_report, "01-analyse-baseline.qmd")`, manually delete the HTML file, and then run `tar_make()`, the newly rendered file is a correct quarto styled output.

Subsequent changes to the report are not being picked up, however: the pipeline runs and finishes, but the document stays how it is.
When I try to force a re-render by deleting the .html, things go sour: 
```
 Error:
! Error running targets::tar_make()
  Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
  Tips: https://books.ropensci.org/targets/debugging.html
  Last error: _paths_ missing files: 01-analyse-baseline.html
```

Which baffles me, since this is the file that I'm trying to create - how can it be required?
In fact, not even the above specified pipeline is reproducible for me because of the error related to the missing file path (it somehow worked the first time, but now I'm simply stuck with the error). I manually deleted the whole `_targets` folder, but the error stays the same.

It might be that I'm not using the package correctly, but this is at least how I'm working with targets and RMarkdown files. Any pointers on how to debug this further would be greatly appreciated!

**Comment by wlandau** [comment]
Sorry, I don't think I can be much help without a reproducible example I can run. All I can really say is that it's a good idea to try to reproduce the error outside `targets`, especially because of https://github.com/quarto-dev/quarto-r/issues/16. I suggest you go through the same steps, but with `quarto::quarto_render("01-analyse-baseline.qmd", execute_dir = getwd(), quiet = FALSE)` instead of `tar_make()`. You may be able to get a more informative error message and resolve the issue outside `targets`/`tarchetypes`.

> Which baffles me, since this is the file that I'm trying to create - how can it be required?

I strongly suspect the error message "_paths_ missing files: 01-analyse-baseline.html" comes from the fact that `1-analyse-baseline.html` was not successfully produced (as opposed to just missing at the beginning, which would indeed have been acceptable).

**Reply by tklebel** [reply]
Thanks for this input - I was able to debug my issue: I relied on `tar_option_set` to load packages for the whole pipeline. However, this seems not to work as expected. When running the quarto doc separately I always added packages (`ggplot2`), as I was aware that this would be necessary. And given that the error message was not forwarded, it was hard to uncover the issue.

I'm not 100% sure what the expected behaviour would be, but I'd expect the following to work, which errors:

```r
targets::tar_dir({  # tar_dir() runs code from a temporary directory.
  lines <- c(
    "---",
    "title: report.qmd source file",
    "output_format: html",
    "---",
    "```{r}",
    "theme_set(theme_bw())",
    "```"
  )
  writeLines(lines, "report.qmd")
  
  targets::tar_script({
    
    targets::tar_option_set(packages = c("tidyverse"))
    
    list(
      tarchetypes::tar_quarto(report, path = "report.qmd")
    )
  }, ask = FALSE)
  targets::tar_make()
})

Error:
! Error running targets::tar_make()
  Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
  Tips: https://books.ropensci.org/targets/debugging.html
  Last error: _paths_ missing files: report.html

```

**Reply by wlandau** [reply]
Quarto runs in its own external process, and  the packages for a target cannot be forwarded to that separate R session. The Quarto document itself is responsible for loading all the packages it needs.

**Reply by tklebel** [reply]
Thanks again. Given that this is different from how RMarkdown works with targets and in particular `tar_option_set`, might it make sense to add some documentation around it? I think it is anyways better practice to explicitly load the packages at the start of the notebook, and once the referenced issue with error propagation is resolved this will be more obvious.

**Comment by tklebel** [comment, marked as answer]
The issue was resolved by loading packages at the top of the quarto document, instead of assuming that `targets::tar_option_set` would do this.
